File: binary-tree/450-delete-from-bst/450.ts
class TreeNode {
	val: number
	left: TreeNode | null
	right: TreeNode | null
	constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
		this.val = (val === undefined ? 0 : val)
		this.left = (left === undefined ? null : left)
		this.right = (right === undefined ? null : right)
	}
}

function deleteNode(root: TreeNode | null, val: number): TreeNode | null {
	function minValueBST(node: TreeNode) {
		while (node.left) {
			node = node.left
		}
		return node
	}

	if (!root) return null;

	// apply binary search
	if (val > root.val) {
		root.right = deleteNode(root.right, val);
	} else if (val < root.val) {
		root.left = deleteNode(root.left, val);
	} else {
		// found the value
		if (!root.left && !root.right) {
			// 0 child case
			return null;
		} else if (!root.left) {
			// 1 child case for the right
			return root.right;
		} else if (!root.right) {
			// 1 child case for the left
			return root.left;
		} else {
			// 2 children case
			let minNode = minValueBST(root.right);
			minNode.left = root.left
			return root.right
		}
	}
	return root;
};

File: binary-tree/450-delete-from-bst/450.go
package binary_tree

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	// to remove a node we must find it first and then
	// decide what to put in place: left or right

	return dfs(root, key)
}

func dfs(node *TreeNode, key int) *TreeNode {
	if node == nil {
		return node
	}

	if key > node.Val {
		node.Right = dfs(node.Right, key)
	} else if key < node.Val {
		node.Left = dfs(node.Left, key)
	} else { // node.Val == key
		// we found our value
		// we need to put the small (left) value on its place
		// and tie the current right value as the left
		if node.Left == nil && node.Right == nil {
			return nil
		} else if node.Right == nil {
			return node.Left
		} else if node.Left == nil {
			return node.Right
		} else {
			// the found node has two children
			node.Val = min(node.Right)
			node.Right = dfs(node.Right, node.Val)
		}
	}
	return node
}

func min(node *TreeNode) int {
	minVal := node.Val

	for node != nil {
		if node.Val < minVal {
			minVal = node.Val
		}
		node = node.Left
	}

	return minVal
}
